ClassNotFoundException when com.datastax.oss:java-driver-mapper-runtime is not in the class path

As of 1.1.0-rc1, CassandraClientProcessor.registerMutinyResultProducerService() throws a ClassNotFoundException when com.datastax.oss:java-driver-mapper-runtime is not in the class path:

```
  @BuildStep
  ServiceProviderBuildItem registerMutinyResultProducerService() {
    return new ServiceProviderBuildItem(
        "com.datastax.oss.driver.api.mapper.result.MapperResultProducerService",
        MutinyResultProducerService.class.getName());
  }
```

The stack trace thrown by `camel-quarkus-integration-test-cassandraql`: 

```
[INFO] --- quarkus-maven-plugin:2.0.0.CR3:build (quarkus-build) @ camel-quarkus-integration-test-cassandraql ---
[INFO] [org.jboss.threads] JBoss Threads version 3.4.0.Final
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  01:02 min
[INFO] Finished at: 2021-06-16T18:39:14+02:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal io.quarkus:quarkus-maven-plugin:2.0.0.CR3:build (quarkus-build) on project camel-quarkus-integration-test-cassandraql: Failed to build quarkus application: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
[ERROR]         [error]: Build step com.datastax.oss.quarkus.deployment.internal.CassandraClientProcessor#registerMutinyResultProducerService threw an exception: java.lang.NoClassDefFoundError: com/datastax/oss/driver/api/mapper/result/MapperResultProducerService
[ERROR]         at java.base/java.lang.ClassLoader.defineClass1(Native Method)
[ERROR]         at java.base/java.lang.ClassLoader.defineClass(ClassLoader.java:1017)
[ERROR]         at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:437)
[ERROR]         at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:397)
[ERROR]         at com.datastax.oss.quarkus.deployment.internal.CassandraClientProcessor.registerMutinyResultProducerService(CassandraClientProcessor.java:409)
[ERROR]         at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[ERROR]         at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
[ERROR]         at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[ERROR]         at java.base/java.lang.reflect.Method.invoke(Method.java:566)
[ERROR]         at io.quarkus.deployment.ExtensionLoader$2.execute(ExtensionLoader.java:820)
[ERROR]         at io.quarkus.builder.BuildContext.run(BuildContext.java:277)
[ERROR]         at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
[ERROR]         at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2442)
[ERROR]         at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1450)
[ERROR]         at java.base/java.lang.Thread.run(Thread.java:829)
[ERROR]         at org.jboss.threads.JBossThread.run(JBossThread.java:501)
[ERROR] Caused by: java.lang.ClassNotFoundException: com.datastax.oss.driver.api.mapper.result.MapperResultProducerService
[ERROR]         at org.codehaus.plexus.classworlds.strategy.SelfFirstStrategy.loadClass(SelfFirstStrategy.java:50)
[ERROR]         at org.codehaus.plexus.classworlds.realm.ClassRealm.unsynchronizedLoadClass(ClassRealm.java:271)
[ERROR]         at org.codehaus.plexus.classworlds.realm.ClassRealm.loadClass(ClassRealm.java:247)
[ERROR]         at org.codehaus.plexus.classworlds.realm.ClassRealm.loadClass(ClassRealm.java:239)
[ERROR]         at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:447)
[ERROR]         at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:397)
[ERROR]         ... 16 more
```

A PR with one possible fix follows.

--- a/deployment/src/main/java/com/datastax/oss/quarkus/deployment/internal/CassandraClientProcessor.java
+++ b/deployment/src/main/java/com/datastax/oss/quarkus/deployment/internal/CassandraClientProcessor.java
@@ -52,7 +52,6 @@ import com.datastax.oss.driver.internal.core.time.ThreadLocalTimestampGenerator;
 import com.datastax.oss.driver.internal.core.tracker.NoopRequestTracker;
 import com.datastax.oss.driver.internal.core.tracker.RequestLogger;
 import com.datastax.oss.quarkus.deployment.api.CassandraClientBuildTimeConfig;
-import com.datastax.oss.quarkus.runtime.internal.mapper.MutinyResultProducerService;
 import com.datastax.oss.quarkus.runtime.internal.quarkus.CassandraClientProducer;
 import com.datastax.oss.quarkus.runtime.internal.quarkus.CassandraClientRecorder;
 import com.datastax.oss.quarkus.runtime.internal.quarkus.CassandraClientStarter;
@@ -408,6 +407,6 @@ class CassandraClientProcessor {
   ServiceProviderBuildItem registerMutinyResultProducerService() {
     return new ServiceProviderBuildItem(
         "com.datastax.oss.driver.api.mapper.result.MapperResultProducerService",
-        MutinyResultProducerService.class.getName());
+        "com.datastax.oss.quarkus.runtime.internal.mapper.MutinyResultProducerService");
   }
 }

--- a/integration-tests/no-mapper/pom.xml
+++ b/integration-tests/no-mapper/pom.xml
@@ -1,0 +1,105 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+
+    Copyright DataStax, Inc.
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+-->
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+  <modelVersion>4.0.0</modelVersion>
+  <parent>
+    <groupId>com.datastax.oss.quarkus</groupId>
+    <artifactId>cassandra-quarkus-integration-tests</artifactId>
+    <version>1.0.3-SNAPSHOT</version>
+  </parent>
+  <artifactId>cassandra-quarkus-integration-tests-no-mapper</artifactId>
+  <name>Cassandra Quarkus :: IT :: No Mapper</name>
+  <dependencies>
+    <!-- Don't depend on cassandra-quarkus-integration-tests-application -->
+    <dependency>
+      <groupId>com.datastax.oss.quarkus</groupId>
+      <artifactId>cassandra-quarkus-client</artifactId>
+    </dependency>
+    <dependency>
+      <groupId>io.quarkus</groupId>
+      <artifactId>quarkus-resteasy</artifactId>
+    </dependency>
+    <dependency>
+      <groupId>io.quarkus</groupId>
+      <artifactId>quarkus-resteasy-jsonb</artifactId>
+    </dependency>
+    <!-- Test dependencies -->
+    <dependency>
+      <groupId>io.quarkus</groupId>
+      <artifactId>quarkus-test-common</artifactId>
+      <scope>test</scope>
+    </dependency>
+    <dependency>
+      <groupId>com.datastax.oss.quarkus</groupId>
+      <artifactId>cassandra-quarkus-test-framework</artifactId>
+      <scope>test</scope>
+    </dependency>
+    <dependency>
+      <groupId>io.quarkus</groupId>
+      <artifactId>quarkus-junit5</artifactId>
+      <scope>test</scope>
+    </dependency>
+    <dependency>
+      <groupId>org.assertj</groupId>
+      <artifactId>assertj-core</artifactId>
+      <scope>test</scope>
+    </dependency>
+    <dependency>
+      <groupId>io.rest-assured</groupId>
+      <artifactId>rest-assured</artifactId>
+      <scope>test</scope>
+    </dependency>
+  </dependencies>
+  <build>
+    <plugins>
+      <plugin>
+        <artifactId>maven-compiler-plugin</artifactId>
+        <!-- No mapper annotation processor -->
+      </plugin>
+      <plugin>
+        <artifactId>maven-jar-plugin</artifactId>
+        <executions>
+          <!--
+            This module only has internal classes, yet we need a javadoc JAR to pass Maven central
+            validation rules.
+          -->
+          <execution>
+            <id>javadoc-jar</id>
+            <phase>package</phase>
+            <goals>
+              <goal>jar</goal>
+            </goals>
+            <configuration>
+              <classifier>javadoc</classifier>
+              <excludes>
+                <exclude>**</exclude>
+              </excludes>
+            </configuration>
+          </execution>
+          <execution>
+            <id>test-jar</id>
+            <goals>
+              <goal>test-jar</goal>
+            </goals>
+          </execution>
+        </executions>
+      </plugin>
+    </plugins>
+  </build>
+</project>

--- a/integration-tests/no-mapper/src/main/java/com/datastax/oss/quarkus/tests/entity/Product.java
+++ b/integration-tests/no-mapper/src/main/java/com/datastax/oss/quarkus/tests/entity/Product.java
@@ -1,0 +1,71 @@
+/*
+ * Copyright DataStax, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.datastax.oss.quarkus.tests.entity;
+
+import java.util.Objects;
+import java.util.UUID;
+
+public class Product {
+
+  private UUID id;
+
+  private String name;
+
+  public Product() {}
+
+  public Product(UUID id, String name) {
+    this.id = id;
+    this.name = name;
+  }
+
+  public UUID getId() {
+    return id;
+  }
+
+  public void setId(UUID id) {
+    this.id = id;
+  }
+
+  public String getName() {
+    return name;
+  }
+
+  public void setName(String name) {
+    this.name = name;
+  }
+
+  @Override
+  public boolean equals(Object o) {
+    if (this == o) {
+      return true;
+    }
+    if (!(o instanceof Product)) {
+      return false;
+    }
+    Product product = (Product) o;
+    return id.equals(product.id) && name.equals(product.name);
+  }
+
+  @Override
+  public int hashCode() {
+    return Objects.hash(id, name);
+  }
+
+  @Override
+  public String toString() {
+    return "Product{id=" + id + ", name='" + name + "'}";
+  }
+}

--- a/integration-tests/no-mapper/src/main/java/com/datastax/oss/quarkus/tests/resource/ProductResource.java
+++ b/integration-tests/no-mapper/src/main/java/com/datastax/oss/quarkus/tests/resource/ProductResource.java
@@ -1,0 +1,67 @@
+/*
+ * Copyright DataStax, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.datastax.oss.quarkus.tests.resource;
+
+import com.datastax.oss.driver.api.core.AsyncPagingIterable;
+import com.datastax.oss.driver.api.core.cql.SimpleStatement;
+import com.datastax.oss.quarkus.runtime.api.session.QuarkusCqlSession;
+import com.datastax.oss.quarkus.tests.entity.Product;
+import java.net.URI;
+import java.util.UUID;
+import java.util.concurrent.CompletionStage;
+import javax.inject.Inject;
+import javax.ws.rs.Consumes;
+import javax.ws.rs.GET;
+import javax.ws.rs.POST;
+import javax.ws.rs.Path;
+import javax.ws.rs.PathParam;
+import javax.ws.rs.Produces;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+import javax.ws.rs.core.Response.Status;
+
+@Path("/product")
+public class ProductResource {
+
+  @Inject QuarkusCqlSession session;
+
+  @POST
+  @Consumes(MediaType.APPLICATION_JSON)
+  public CompletionStage<Response> createProduct(Product product) {
+    return session
+        .executeAsync(
+            SimpleStatement.newInstance(
+                "INSERT INTO product (id, name) VALUES (?, ?)", product.getId(), product.getName()))
+        .thenApply((ignore) -> Response.created(URI.create("/product/" + product.getId())).build());
+  }
+
+  @GET
+  @Path("/{id}")
+  @Produces(MediaType.APPLICATION_JSON)
+  public CompletionStage<Response> getProduct(@PathParam("id") UUID id) {
+    return session
+        .executeAsync(SimpleStatement.newInstance("SELECT id, name FROM product WHERE id = ?", id))
+        .thenApply(AsyncPagingIterable::one)
+        .thenApply(
+            row -> {
+              if (row == null) {
+                return Response.status(Status.NOT_FOUND).build();
+              } else {
+                return Response.ok(new Product(row.getUuid(0), row.getString(1))).build();
+              }
+            });
+  }
+}

--- a/integration-tests/no-mapper/src/main/resources/application.conf
+++ b/integration-tests/no-mapper/src/main/resources/application.conf
@@ -1,0 +1,43 @@
+# Configuration overrides for integration tests.
+# The settings below are mostly related to timeouts and provide higher values
+# suitable for running integration tests in CI environments.
+# Note: we don't have integration tests for Graph or Continuous Paging yet,
+# but for consistency, their relevant timeouts are being modified to be similar
+# to other timeout values below.
+datastax-java-driver {
+  basic {
+    request.timeout = 10 seconds
+    graph.timeout = 10 seconds
+  }
+  advanced {
+    connection {
+      connect-timeout = 10 seconds
+      init-query-timeout = 10 seconds
+      set-keyspace-timeout = 10 seconds
+    }
+    heartbeat.timeout = 10 seconds
+    control-connection.timeout = 10 seconds
+    control-connection.schema-agreement.timeout = 10 seconds
+    metadata.schema.request-timeout = 10 seconds
+    prepared-statements.reprepare-on-up.timeout = 10 seconds
+    continuous-paging.timeout {
+      first-page = 10 seconds
+      other-pages = 10 seconds
+    }
+    metrics {
+      // Raise histogram bounds because the tests execute DDL queries with a higher timeout
+      session.cql_requests.highest_latency = 15 seconds
+      session.graph-requests.highest-latency = 15 seconds
+      node.cql_messages.highest_latency = 15 seconds
+    }
+    // adjust quiet period to 0 seconds to speed up tests
+    netty {
+      io-group {
+        shutdown {quiet-period = 0, timeout = 15, unit = SECONDS}
+      }
+      admin-group {
+        shutdown {quiet-period = 0, timeout = 15, unit = SECONDS}
+      }
+    }
+  }
+}

--- a/integration-tests/no-mapper/src/main/resources/application.properties
+++ b/integration-tests/no-mapper/src/main/resources/application.properties
@@ -1,0 +1,25 @@
+#
+# Copyright DataStax, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+quarkus.cassandra.contact-points=${quarkus.cassandra.docker_host}:${quarkus.cassandra.docker_port}
+quarkus.cassandra.local-datacenter=datacenter1
+quarkus.cassandra.keyspace=ks1
+
+#quarkus.log.level=DEBUG
+#quarkus.log.min-level=DEBUG
+#quarkus.log.category."com.datastax.oss.quarkus".level=DEBUG
+#quarkus.log.category."com.datastax.oss.driver".level=DEBUG
+#quarkus.log.category."com.datastax.dse.driver".level=INFO

--- a/integration-tests/no-mapper/src/main/resources/init_script.cql
+++ b/integration-tests/no-mapper/src/main/resources/init_script.cql
@@ -1,0 +1,8 @@
+CREATE KEYSPACE IF NOT EXISTS ks1
+    WITH replication = {'class':'SimpleStrategy', 'replication_factor':1};
+
+CREATE TABLE IF NOT EXISTS ks1.product
+(
+    id   uuid PRIMARY KEY,
+    name text
+);

--- a/integration-tests/no-mapper/src/test/java/com/datastax/oss/quarkus/tests/ProductResourceIT.java
+++ b/integration-tests/no-mapper/src/test/java/com/datastax/oss/quarkus/tests/ProductResourceIT.java
@@ -1,0 +1,58 @@
+/*
+ * Copyright DataStax, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.datastax.oss.quarkus.tests;
+
+import static io.restassured.RestAssured.given;
+import static io.restassured.RestAssured.when;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.hamcrest.Matchers.endsWith;
+
+import com.datastax.oss.quarkus.test.CassandraTestResource;
+import com.datastax.oss.quarkus.tests.entity.Product;
+import io.quarkus.test.common.QuarkusTestResource;
+import io.quarkus.test.junit.QuarkusTest;
+import io.restassured.http.ContentType;
+import java.util.UUID;
+import javax.ws.rs.core.Response.Status;
+import org.junit.jupiter.api.Test;
+
+@QuarkusTest
+@QuarkusTestResource(CassandraTestResource.class)
+public class ProductResourceIT {
+
+  @Test
+  public void should_create_and_retrieve_product() {
+    Product expected = new Product(UUID.randomUUID(), "product1");
+    given()
+        .body(expected)
+        .contentType(ContentType.JSON)
+        .when()
+        .post("/product")
+        .then()
+        .statusCode(Status.CREATED.getStatusCode())
+        .header("location", endsWith("/product/" + expected.getId()));
+    Product actual =
+        when()
+            .get("/product/{id}", expected.getId())
+            .then()
+            .contentType(ContentType.JSON)
+            .statusCode(Status.OK.getStatusCode())
+            .extract()
+            .body()
+            .as(Product.class);
+    assertThat(actual).isEqualTo(expected);
+  }
+}

--- a/integration-tests/pom.xml
+++ b/integration-tests/pom.xml
@@ -34,6 +34,7 @@
     <module>main</module>
     <module>metrics-microprofile</module>
     <module>metrics-disabled</module>
+    <module>no-mapper</module>
   </modules>
   <dependencyManagement>
     <dependencies>

